File: src/io/github/hulang1024/chinesechess/rule/chess/ChessN.java
package io.github.hulang1024.chinesechess.rule.chess;

import io.github.hulang1024.chinesechess.rule.AbstractChess;
import io.github.hulang1024.chinesechess.rule.ChessPosition;
import io.github.hulang1024.chinesechess.rule.ChineseChessGame;

/**
 * 马
 * @author Hu Lang
 */
public class ChessN extends AbstractChess {
    @Override
    public boolean canGoTo(ChessPosition destPos, ChineseChessGame game) {
        // 马走“日”，蹩马腿
        int rowOffset = destPos.row - pos.row;
        int colOffset = destPos.col - pos.col;

        if (Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 1) {
            return game.chessboard.isEmpty(pos.row + (rowOffset > 0 ? +1 : -1), pos.col);
        } else if (Math.abs(rowOffset) == 1 && Math.abs(colOffset) == 2) {
            return game.chessboard.isEmpty(pos.row, pos.col + (colOffset > 0 ? +1 : -1));
        } else {
            return false;
        }
    }
}

File: web-client/chinese-chess/src/rule/chess/ChessN.ts
import ChessHost from "../chess_host";
import ChessPos from "../ChessPos";
import Game from "../Game";
import AbstractChess from "./AbstractChess";
import { sign } from "./move_rules";

/**
 * 马
 */
export default class ChessN extends AbstractChess {
    constructor(pos: ChessPos, host: ChessHost) {
        super(pos, host);
    }

    canGoTo(destPos: ChessPos, game: Game) {
        // 马走“日”，蹩马腿
        const rowOffset = destPos.row - this.pos.row;
        const colOffset = destPos.col - this.pos.col;

        if (Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 1) {
            return game.getChessboard().isEmpty(this.pos.row + sign(rowOffset), this.pos.col);
        } else if (Math.abs(rowOffset) == 1 && Math.abs(colOffset) == 2) {
            return game.getChessboard().isEmpty(this.pos.row, this.pos.col + sign(colOffset));
        } else {
            return false;
        }
    }
}
